pkg/process/util/api/weighted_queue.go

```diff
@@ -0,0 +1,208 @@
+	return &WeightedQueue{
+		dataAvailable: make(chan struct{}),
+		queue:         list.New(),
+		maxSize:       maxSize,
+		maxWeight:     maxWeight,
+		currentWeight: 0,
+	}
+}
+
+// Len returns the number of items in the queue
+func (q *WeightedQueue) Len() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Len()
+}
+
+// Weight returns the current weight of the queue
+func (q *WeightedQueue) Weight() int64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.currentWeight
+}
+
+// Poll retrieves the head of the queue or blocks until an item is available.  The provided exit channel can be closed
+// to interrupt the blocking operation.  Returns the head of the queue and true or nil, false if the poll was
+// interrupted by the closing of the exit channel
+func (q *WeightedQueue) Poll(exit chan struct{}) (WeightedItem, bool) {
+	for {
+		select {
+		case <-exit:
+			return nil, false
+		default:
+
+		}
+
+		q.mu.Lock()
+
+		// If the queue is empty, wait for a signal on the dataAvailable channel
+		if q.queue.Len() == 0 {
+			q.mu.Unlock()
+
+			select {
+			case <-q.dataAvailable:
+				continue
+			case <-exit:
+				return nil, false
+			}
+		}
+
+		e := q.queue.Front()
+		item := e.Value.(WeightedItem)
+		q.queue.Remove(e)
+		q.currentWeight -= item.Weight()
+
+		q.mu.Unlock()
+
+		return item, true
+	}
+}
+
+// Add adds the item to the queue.
+func (q *WeightedQueue) Add(item WeightedItem) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// If the item won't fit, don't even bother trying
+	if item.Weight() > q.maxWeight {
+		return
+	}
+
+	newWeight := q.currentWeight + item.Weight()
+
+	if newWeight > q.maxWeight {
+		weightToRemove := newWeight - q.maxWeight
+		weightRemoved := int64(0)
+
+		// Try to find an item of the same type that we can expire
+		for iter := q.iterator(); iter.hasNext(); iter.next() {
+			if v := iter.value(); v.Type() == item.Type() {
+				iter.remove()
+				weightRemoved += v.Weight()
+				if weightRemoved >= weightToRemove {
+					break
+				}
+			}
+		}
+
+		// If we didn't find enough free weight removing similar items, start purging the earliest items
+		// until there is room
+		if weightRemoved < weightToRemove {
+			for iter := q.iterator(); iter.hasNext(); iter.next() {
+				v := iter.value()
+				iter.remove()
+				weightRemoved += v.Weight()
+				if weightRemoved >= weightToRemove {
+					break
+				}
+			}
+		}
+	}
+
+	// If the queue is full, expire a single item to make room
+	if q.queue.Len() == q.maxSize {
+		// Try to find an item of the same type that we can expire
+		removed := false
+		for iter := q.iterator(); iter.hasNext(); iter.next() {
+			if v := iter.value(); v.Type() == item.Type() {
+				iter.remove()
+				removed = true
+				break
+			}
+		}
+
+		// No similar items, remove the oldest element from the queue
+		if !removed {
+			q.queue.Remove(q.queue.Front())
+		}
+	}
+
+	q.queue.PushBack(item)
+	q.currentWeight = newWeight
```

Comment:
It seems to me `newWeight` doesn't account for all the operations you did along the way.

Reply:
Yes, thank you.  Fixed.